[TFLite] Refactor and centralize LUT generation and access code into headers

Tessil opened the pull request:
Hello,

This PR refactors and centralizes the `PopulateLookupTable` and `gen_lut` functions into a single `LUTPopulate` function that can generate uint8, int8 and int16 LUTs and be used by multiple operators.

The change is done in multiple steps:
* The int8/uint8 `PopulateLookupTable` from `tensorflow/lite/kernels/activations.cc` is moved into `tensorflow/lite/kernels/internal/common.h` with a slight signature change and `gen_lut` is renamed to `PopulateLookupTable<int16_t>` with also a slight signature change. This change centralizes all the LUT generation in a single templated `PopulateLookupTable<T>` function.
* The `std::function` of the int8/uint8 `PopulateLookupTable` is replaced by a raw function pointer so that it can be used in TFLite Micro.
* The code used for `gen_lut` is now only used for int16 -> int16 LUT generation. int8 -> int8, int8 -> int16 and int16 -> int8 support was removed as unused and not fundamentally necessary.
* All the LUT generation calls are adapted to these changes.
* We made sure that the change is numerically backward compatible, all the generated LUTs have the exact same values.
* The LUT lookup code was also moved into separate headers `tensorflow/lite/kernels/internal/optimized/integer_ops/lut.h` and `tensorflow/lite/kernels/internal/reference/integer_ops/lut.h` to centralize the LUT access solving the `// TODO(b/143696793): move this to optimized_ops.` comment.
* `gen_lut/PopulateLookupTable`, `lut_size` and `lut_lookup` were renamed to `LUTPopulate`, `LUTSize` and `LUTLookup` respectively in a separate commit. This change is done to have more uniform names for these related functions and be consistent with the CamelCase function naming used in the codebase. This separate commit is minor and optional and could eventually be removed if deemed unnecessary.

These modifications don't change the numerical behaviour of the operators and don't add any new feature. They are made to avoid any potential future conflicts and confusions between the `PopulateLookupTable` and `lut_gen` functions and make the LUT generation code accessible to other operators (example of a [potential EXP quantization PR](https://github.com/Tessil/tensorflow/commit/b12f4980e26a0a79bc0019f6965c518d24062f57) that is based on these changes).

Thibaut

Tessil committed:
Refactor and centralize the LUT generation and access code

Tessil committed:
Rename LUT generation and access functions for coherence

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/refactor_lut_generation

abattery commented on tensorflow/lite/kernels/internal/common.h:
```diff
@@ -375,55 +375,80 @@ inline Integer FloorLog2(Integer n) {
+                    std::is_same<T, int16_t>::value,
+                "Only LUTs with uint8, int8 or int16 inputs are supported.");
+  if (std::is_same<T, uint8_t>::value || std::is_same<T, int8_t>::value) {
+    return 256;
```
could you add a comment for describing the meaning of this number?

Tessil replied:
The numbers are described in the comment of the LUTSize function. I extended it a bit.

abattery commented on tensorflow/lite/kernels/internal/common.h:
```diff
@@ -375,55 +375,80 @@ inline Integer FloorLog2(Integer n) {
+  if (std::is_same<T, uint8_t>::value || std::is_same<T, int8_t>::value) {
+    return 256;
+  } else {
+    return 513;
```
could you add a comment for describing the meaning of this number?

Tessil replied:
See above

abattery commented on tensorflow/lite/kernels/internal/reference/integer_ops/lut.h:
```diff
@@ -0,0 +1,34 @@
+template <typename InputT, typename OutputT>
+inline void LookupTable(const InputT* input_data, int num_elements,
+                        const OutputT* lut, OutputT* output_data) {
+  for (int i = 0; i < num_elements; i++) {
```
++i

Tessil replied:
Done

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/refactor_lut_generation

abattery reviewed (changes requested):
Please resolve the comments.

Tessil committed:
Extend the comment regarding the LUTSize()

Tessil committed:
Use ++i instead of i++

Tessil committed:
Fix Windows build for GELU LUT generation

Tessil commented:
@abattery I fixed a compilation error with MSVC in the latest commit, the PR will need re-approval. Thanks!

abattery commented:
Looks like it does not build well..

abattery reviewed (commented):
Currently, looks like it is not buildable well..

Tessil commented:
What is the error? The Linux GitHub CI seems to work well while the Windows ones failing seem to be from an unrelated error.

Is it tflite-micro that fails? I realize that https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/kernels/softmax_common.cc#L58 must be changed in the same way as https://github.com/tensorflow/tensorflow/pull/54390/files#diff-82b5be35a42ea52dba4ecf726fe23e742054d53f514803d6b5f48b8da60b8828R591 but using `float` instead of `double`. We can eventually create a PR in tflite-micro if it's easier for you.

abattery commented:
For example,

```
[lut.h:47]:25: error: use of undeclared identifier 'optimized_ops'
    uint8x16_t output = optimized_ops::aarch64_lookup_vector(table, input);
                        ^
1 error generated.
```

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/refactor_lut_generation

Tessil committed:
Add missing include for aarch64_lookup_vector

Tessil commented:
Effectively, I fixed the missing include. Thanks!

abattery reviewed (changes requested):
The PR isn't buildable.

[tensorflow/lite/kernels/table.cc:17 //tensorflow/lite/kernels:custom_ops does not depend on a module exporting ...
'tensorflow/lite/kernels/internal/reference/integer_ops/lut.h'
#include "tensorflow/lite/kernels/internal/reference/integer_ops/lut.h"

1 error generated.

Tessil committed:
Add missing dependency to 'custom_ops' target

Tessil commented:
Thanks, last commit should hopefully fix it (though difficult to check as the GitHub CI doesn't show the error).

abattery reviewed (changes requested):
error: use of undeclared identifier 'gen_lut'
    gen_lut<float, int16_t, int16_t>(
    ^
[tflite_micro/tensorflow/lite/micro/kernels/softmax_common.cc:63]

Still, it is not buildable

Tessil commented:
Yes, the PR requires some changes in tflite-micro as I mentioned in my https://github.com/tensorflow/tensorflow/pull/54390#issuecomment-1156286087 comment that I can't do in this PR.

What would be the best way? Could you do it internally? Or should I raise a PR in the tflite-micro repository?

abattery commented:
Enabling both at once is really hard. You might want to land the new features without enabling by default and then you might be able to carefully turn on in the separate PRs for each repository.

Tessil committed:
Keep a gen_lut function to be backward-compatible with tflite-micro.

This gen_lut function will be removed once its usage in tflite-micro is replaced with LUTPopulate.

Tessil commented:
Thanks, I kept a restricted `gen_lut` function to be compatible with the two calls in https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/kernels/softmax_common.cc#L58 

Once merged, we should then modify tflite-micro to use `LUTPopulate` instead and removed this temporary `gen_lut`.

abattery commented:
Could you sync again? The current sync head wasn't green enough.

Tessil commented:
Sorry, what do you mean by sync? Should I rebase my PR? Thanks

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/refactor_lut_generation

Tessil commented:
Thanks, I merged the latest changes (if it what was meant by sync).

abattery commented:
I don't have permissions to approve the TOSA related changes. I will manually merge this change without the TOSA comment updates. Could you send a separate change for the TOSA related one and get approvals from the right owner instead?

Tessil commented:
@abattery Thanks, I'll check to create a new PR for the TOSA comment change. It seems though that the PR was rolled back, was there a problem with an internal test?

alkejah commented:
Good day sir, I'm from Philippines we found your activities are helpful, sir we are developing a character recognition system but we are struggling on how to make it an android app and we are also struggling if we are doing a right thing with our thesis project, hope you can help us sir, we are willing to send you the code sir if you want to see and inspect, thank you in advance sir, it would be greate help to us. Thank you sir.
